File: 931. Minimum Falling Path Sum/minFallingPathSum.cpp

class Solution
{
public:
    int solve(vector<vector<int>> &A, int row, int col, vector<vector<int>> &dp)
    {
        int n = A.size();

        if (dp[row][col] != INT_MAX)
            return dp[row][col];

        if (row == n - 1)
            return dp[row][col] = A[row][col];

        int left = INT_MAX, right = INT_MAX;

        if (col > 0)
            left = solve(A, row + 1, col - 1, dp);

        int straight = solve(A, row + 1, col, dp);

        if (col < n - 1)
            right = solve(A, row + 1, col + 1, dp);

        dp[row][col] = min(left, min(straight, right)) + A[row][col];

        return dp[row][col];
    }
    int minFallingPathSum(vector<vector<int>> &matrix)
    {

        int n = matrix.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, INT_MAX));
        int ans = INT_MAX;
        for (int i = 0; i < n; ++i)
        {
            ans = min(ans, solve(matrix, 0, i, dp));
        }

        return ans;
    }
};

File: 931. Minimum Falling Path Sum/minFallingPathSum.java
class Solution {
    public int solve(int[][] A, int row, int col, int[][] dp) {
        int n = A.length;

        if (dp[row][col] != Integer.MAX_VALUE)
            return dp[row][col];

        if (row == n - 1) {
            return dp[row][col] = A[row][col];
        }

        int left = Integer.MAX_VALUE, right = Integer.MAX_VALUE;

        if (col > 0) {
            left = solve(A, row + 1, col - 1, dp);
        }

        int straight = solve(A, row + 1, col, dp);

        if (col < n - 1) {
            right = solve(A, row + 1, col + 1, dp);
        }

        dp[row][col] = Math.min(left, Math.min(straight, right)) + A[row][col];

        return dp[row][col];
    }

    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int[][] dp = new int[n + 1][n + 1];
        for (int[] row : dp) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }

        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < n; ++i) {
            ans = Math.min(ans, solve(matrix, 0, i, dp));
        }

        return ans;
    }
}